torch/csrc/utils/python_numbers.h

```diff
@@ -126,3 +126,12 @@ inline double THPUtils_unpackDouble(PyObject* obj) {
   }
   return value;
 }
+
+inline std::complex<double> THPUtils_unpackComplexDouble(PyObject *obj) {
+  Py_complex value = PyComplex_AsCComplex(obj);
+  if (value.real == -1.0 && PyErr_Occurred()) {
```

Comment:
Why do you test if `value.real == -1.0`? Wouldn't it be better to test `PyComplex_Check` first?

Reply:
Err... yes, I missed `PyComplex_Check`, but I think we will still need that `value.real == -1.0` according to the doc of `PyComplex_AsCComplex`, it will try `__complex__` for objects that defines this magic method and return a real number `-1.0` if this fails. Ref: https://docs.python.org/3/c-api/complex.html

And similar with pybind11: https://github.com/pybind/pybind11/blob/master/include/pybind11/complex.h

I'll add a `PyComplex_Check` anyway.

Reply:
Asked pybind11 people here: https://github.com/pybind/pybind11/issues/1518

Reply:
Per https://github.com/pybind/pybind11/commit/259b2fafea69c546111f1eab814e46afed17a1e6 we concluded that we should check for `-1.0`.